File: Lectures/Algorithms-Analysis/Egyptian-Fraction.cpp
#include <bits/stdc++.h>
using namespace std;

void egyptian_fraction(int nr, int dr) {
    vector<int> ef;
    while (nr != 0) {
        int x = (dr + (nr - 1)) / nr;
        ef.push_back(x);
        nr = x * nr - dr;
        dr = dr * x;
    }
    for (int i = 0; i < ef.size(); i++) {
        printf("1/%d", ef[i]);
        if (i < ef.size() - 1)
            printf(" + ");
    }
}

int main() {
    int nr = 6, dr = 14;
    printf("The Egyptian Fraction Representation of %d/%d is\n", nr, dr);
    egyptian_fraction(nr, dr);
}

File: Lectures/Algorithms-Analysis/Egyptian-Fraction.py
def egyptian_fraction(nr, dr):
    ef = []
    while nr != 0:
        x = (dr + (nr-1)) // nr
        ef.append(x)
        nr = x * nr - dr
        dr = dr * x
    for i in range(len(ef)):
        print("1/%d" % ef[i], end = '')
        if i < len(ef) - 1:
            print(' + ', end='')

nr, dr = 6, 14
print("The Egyptian Fraction Representation of %d/%d is" % (nr, dr))
egyptian_fraction(nr, dr)
